localstack/services/secretsmanager/provider.py

```diff
@@ -0,0 +1,352 @@
+    AddReplicaRegionListType,
+    BooleanType,
+    CancelRotateSecretResponse,
+    ClientRequestTokenType,
+    CreateSecretResponse,
+    DeleteResourcePolicyResponse,
+    DeleteSecretResponse,
+    DescribeSecretResponse,
+    DescriptionType,
+    GetResourcePolicyResponse,
+    GetSecretValueResponse,
+    KmsKeyIdType,
+    ListSecretVersionIdsResponse,
+    MaxResultsType,
+    NameType,
+    NextTokenType,
+    NonEmptyResourcePolicyType,
+    PutResourcePolicyResponse,
+    PutSecretValueResponse,
+    RecoveryWindowInDaysType,
+    RemoveRegionsFromReplicationResponse,
+    RemoveReplicaRegionListType,
+    ReplicateSecretToRegionsResponse,
+    RestoreSecretResponse,
+    RotateSecretResponse,
+    RotationLambdaARNType,
+    RotationRulesType,
+    SecretBinaryType,
+    SecretIdType,
+    SecretsmanagerApi,
+    SecretStringType,
+    SecretVersionIdType,
+    SecretVersionStagesType,
+    SecretVersionStageType,
+    StopReplicationToReplicaResponse,
+    TagKeyListType,
+    TagListType,
+    UpdateSecretResponse,
+    UpdateSecretVersionStageResponse,
+    ValidateResourcePolicyResponse,
+)
+from localstack.constants import TEST_AWS_ACCOUNT_ID
+from localstack.services.moto import call_moto, call_moto_with_request
+from localstack.utils.aws import aws_stack
+from localstack.utils.common import to_str
+from localstack.utils.strings import short_uid
+
+LOG = logging.getLogger(__name__)
+
+
+class SecretsmanagerProvider(SecretsmanagerApi, ABC):
+    def __init__(self):
+        super().__init__()
+        apply_patches()
+
+    @staticmethod
+    def _transform_context_secret_id(context: RequestContext) -> Optional[Dict]:
+        # If secret ARN ends with "-<randomId>" this is removed from the request for upstream compatibility.
+        data = json.loads(to_str(context.request.data or "{}"))
+        secret_id = data.get("SecretId", None)
+        if secret_id and ":" in secret_id:
+            arn = aws_stack.parse_arn(secret_id)
+            aws_region = aws_stack.get_region()
+            if arn["region"] != aws_region:
+                LOG.info(f'Expected request region "{aws_region}" for secret "{secret_id}"')
+            resource_id = arn["resource"].split(":")[-1]
+            if resource_id[-7] == "-":
+                data["SecretId"] = resource_id[:-7]
+            elif resource_id[-1] != "-":
+                data["SecretId"] += "-"
+            return data
+        return None
+
+    @staticmethod
+    def _call_moto_with_request_secret_id(context: RequestContext) -> ServiceResponse:
+        data_dict = SecretsmanagerProvider._transform_context_secret_id(context)
+        return call_moto_with_request(context, data_dict) if data_dict else call_moto(context)
+
+    def cancel_rotate_secret(
+        self, context: RequestContext, secret_id: SecretIdType
+    ) -> CancelRotateSecretResponse:
+        return self._call_moto_with_request_secret_id(context)
+
+    def create_secret(
```

Comment:
I have added this to all similar scenarios except `PutResourcePolicy` and `ValidateResourcePolicy, as if non expanded requests never reach the provider. I don't know if we want to postpone this change and investigate why this might be.

Reply:
if we add `expand=False`, then the method signature should have only two arguments: the `RequestContext`, and the input shape of the operation. most of the time it will be something like `<Operation>Request` or <Operation>Input` so `CreateSecretRequest`. The signature of the method should therefore be:
`def create_secret(context: RequestContext, request: CreateSecretRequest)` this is what `expand=False` does

Reply:
Thanks for the overview, this indeed appears to have resolved that behaviour I mentioned above!

Reply:
Thanks for the overview, this indeed appears to have resolved that behaviour I mentioned above!